### DISCUSSION TITLE
`content script` inject

### DISCUSSION THREAD

**Discussion by xunziheng** [wxt-dev/wxt#757] [Q&A] [answered]
For example:

I am using `content script` to inject scripts into the specified matching page, such as `https://test.local/<ids>/edit`,

but when the user visits the website, he first enters `https://test.local/<ids>/view`,

and then the user clicks the `edit` button, the URL changes to `.../edit`.

This is a simple change, but I observe that the content script is not injected.

It is injected only when I refresh the page manually.

How can I inject these scripts when the user enters the matching page?

Waiting for your expert advice. @aklinker1

**Comment by aklinker1** [comment]
Seems like the site you're working with is a SPA, or the website uses HTML 5 history to switch pages without a full reload. In this case, you need to load your content script for the whole site, not just the `/edit` path, and track the page changes yourself, deciding when to execute your logic yourself.

Thankfully, WXT provides utils for this.

```ts
export default defineContentScript({
  matches: ["*://test.local/<id>/*"],
  main(ctx) {
    const editPattern = new MatchPattern("*://test.local/<id>/edit");

    if (editPattern.matches(location)) editMain(ctx);

    ctx.addEventListener(window, "wxt:locationchange", (event) => {
      if (editPattern.matches(event.newUrl) && !editPattern.matches(event.oldUrl)) 
        editMain(ctx);
    })
  }
})

function editMain(ctx: ContentScriptContext) {
  // Your "/edit" code here
}
```

> Wrote this on my phone by memory, so some of the APIs might be slightly wrong, misspelled, or missing curly braces. But hopefully it gives you an idea of how to accomplish this.

**Reply by aklinker1** [reply]
Letting WXT write this boilerplate would be a good feature idea. Something like `spa: true`... I'll think about it

**Reply by xunziheng** [reply]
sorry I tried that but not work:

(yes I am glad to see your great idea maybe like `spa: true`)

```js
export default defineContentScript({
  matches: ['<all_urls>'],

  main(ctx) {
    ctx.addEventListener(document, "wxt:locationchange", (event) => {
      console.log('teststest', event)
    })
  },
});
```

**Reply by xunziheng** [reply]
by the way I am using `wxt@0.18.7`

I got a spa app, when I toggle the page, it was nothing go.

**Reply by aklinker1** [reply, marked as answer]
Try `window` instead of `document`:

https://github.com/wxt-dev/wxt/blob/394cb1a52051d210e0685befb3ca9324778b1aa7/packages/wxt-demo/src/entrypoints/location-change.content.ts#L6-L8

**Reply by xunziheng** [reply]
well done, thanks, but can you show me why?

**Reply by aklinker1** [reply]
https://github.com/wxt-dev/wxt/blob/9fc6408ef9c82a7a60268fedf5619a5ee719ed39/packages/wxt/src/client/content-scripts/location-watcher.ts#L24

I call `window.dispatchEvent`, not `document.dispatchEvent`
